### PR TITLE
Remove unnecessary hack which causes data races

### PR DESCRIPTION
**Pull request by yashykt**
Fixes P1 bug #16369. The issue is explained in the bug.
Remove unnecessary hack which causes data races and eventual segfaults.

### CODE REVIEW

File: src/core/ext/transport/chttp2/transport/chttp2_transport.cc [hunk 1]
```diff
@@ -1409,21 +1407,13 @@ static void perform_stream_op_locked(void* stream_op,
...
-  // TODO(roth): This is a hack needed because we use data inside of the
-  // closure itself to do the barrier calculation (i.e., to ensure that
-  // we don't schedule the closure until all ops in the batch have been
-  // completed).  This can go away once we move to a new C++ closure API
-  // that provides the ability to create a barrier closure.
-  if (on_complete == nullptr) {
-    on_complete = GRPC_CLOSURE_INIT(&op->handler_private.closure, do_nothing,
-                                    nullptr, grpc_schedule_on_exec_ctx);
+  if (on_complete != nullptr) {
+    /* This batch has send ops. Use final_data as a barrier until enqueue time;
```
Comment (vjpai) [inline src/core/ext/transport/chttp2/transport/chttp2_transport.cc]: Can you switch to C++ comments for consistency

Comment (vjpai) [reply src/core/ext/transport/chttp2/transport/chttp2_transport.cc]: Can you also add a comment above the if statement explaining the circumstances under which on_complete would either be nullptr or not

Comment (yashykt) [reply src/core/ext/transport/chttp2/transport/chttp2_transport.cc]: done!

**Comment by yashykt** [comment]
Test failure should get fixed by #18475

**Comment by yashykt** [comment]
I let the test run for some time and did not see it fail, compared to earlier where it would fail faster.

**Comment by yashykt** [comment]
Thanks for reviewing!

**Pull request merged**